README
.. -*- coding: utf-8 -*-

==================================================
   pyofa: Python Wrapper for MusicDNS  (libofa)
==================================================

.. contents::
..
    1  Description
    2  Dependencies
      2.1  MusicDNS Key
    3  Documentation
    4  Download
    5  Build and Installation
      5.1  Installing on Ubuntu
    6  Copyright and License
    7  Author

Description
===========

pyofa is a simple Python wrapper around the MusicDNS technology
provided by MusicIP for generating PUIDs (audio fingerprints) from
audio files (e.g. MP3, M4A, etc.).


Dependencies
============

- `Python 2.5 <http://python.org>`_ or higher
- MusicIP's `libofa <http://www.musicip.com/dns>`_
- python-mutagen
- libavcodec (FFMPEG)

MusicDNS Key
------------

You will need your own MusicDNS key to use this package. You simply
provide it to the `getpuid()` function. You register for a key here:

  http://www.musicip.com/dns/key.jsp


Documentation
=============

- See `example script <bin/musicdns-getpuid>`_
- `CHANGES <CHANGES>`_
- `TODO <TODO>`_


Download
========

- `Download </downloads/pyofa/>`_

For the absolute latest version, get it from the Mercurial
repository::

  hg clone https://hg.furius.ca/public/pyofa



Build and Installation
======================

To build the dynlibs::

  python setup.py config

In order for this to work, it should say "yes" for both ofa and
avcodec. Then, to build and install::

  python setup.py build_ext --inplace
  python setup.py install 

or to build an RPM (with AVCodec/FFMPEG and OFA enabled)::

  python setup.py bdist_rpm


Installing on Ubuntu
--------------------
Ubuntu, here are the list of packages you need to install::

  # apt-get install python-mutagen libavcodec-dev libavformat-dev libofa0-dev


Copyright and License
=====================

This code is distributed under the `GNU General Public License <COPYING>`_;


Author
======

Martin Blais <blais@furius.ca>
(using the code of Lukáš Lalinský)

.. note::

   Lukáš is the author of the great majority of the code, I just
   repackaged and simplified it outside of Picard because I wanted to
   write scripts using MusicDNS without a GUI application.





CHANGES
====================
   pyofa: CHANGES
====================


Version 1.0.3
=============

* RPM configuration to build directly as bdist_rpm without config step
* WAV file duration cannot be determined by mutagen, papered over this


Version 1.0.2
=============

* Fixed bugs with music-guess-album and with import errors.


Version 1.0.1
=============

* Backported to 2.4.


Version 1.0
===========

* Initial release.


old/findpuid.py
#! /usr/bin/env python
# 
# Search for a track by title (and optionally by artist name).
#
# Usage:
#   python findtrack.py 'track name' ['artist name']
#
# $Id$
#
import sys
import logging
from musicbrainz2.webservice import Query, TrackFilter, WebServiceError

logging.basicConfig()
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)
    

if len(sys.argv) < 1:
    print "Usage: findtrack.py 'puid'"
    sys.exit(1)

q = Query()
try:
    f = TrackFilter(puid=sys.argv[1])
    results = q.getTracks(f)
except WebServiceError, e:
    print 'Error:', e
    sys.exit(1)


for result in results:
    track = result.track
    print "Score     :", result.score
    print "Id        :", track.id
    print "Title     :", track.title
    print "Artist    :", track.artist.name
    print

# EOF




#!/usr/bin/env python
"""
Find a track's artist and title.
"""

def find_artist_and_title(puid):
    q = Query()
    try:
        f = TrackFilter(puid=sys.argv[1])
        results = q.getTracks(f)
    except WebServiceError, e:
        print 'Error:', e
        sys.exit(1)
    

def main():
    import optparse
    parser = optparse.OptionParser(__doc__.strip())
    opts, args = parser.parse_args()

    for fn in args:
        



if __name__ == '__main__':
    main()


old/musicdnsws.py
# -*- coding: utf-8 -*-
#
# Picard, the next-generation MusicBrainz tagger
# Copyright (C) 2006 Lukáš Lalinský
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

import musicbrainz2.webservice
import urllib
from musicbrainz2.webservice import ResponseError
from musicbrainz2.wsxml import MbXmlParser, ParseError


class TrackFilter(musicbrainz2.webservice.IFilter):

    def __init__(self, clientId=None, clientVersion=None, fingerprint=None,
                 puid=None, fileSha1=None, acousticSha1=None, metadata=None,
                 bitrate=None, format=None, length=None, artist=None,
                 title=None, album=None, trackNum=None, genre=None, year=None,
                 composer=None, conductor=None, orchestra=None, encoding=None,
                 lookupType=None):
        self._params = [
            ('lkt', lookupType),
            ('cid', clientId),
            ('cvr', clientVersion),
            ('fpt', fingerprint),
            ('uid', puid),
            ('s1f', fileSha1),
            ('s1a', acousticSha1),
            ('rmd', metadata),
            ('brt', bitrate),
            ('fmt', format),
            ('dur', length),
            ('art', artist),
            ('ttl', title),
            ('alb', album),
            ('tnm', trackNum),
            ('gnr', genre),
            ('yrr', year),
            ('cmp', composer),
            ('cnd', conductor),
            ('orc', orchestra),
            ('enc', encoding),
        ]

    def createParameters(self):
        return musicbrainz2.webservice._createParameters(self._params)


class Query(musicbrainz2.webservice.Query):

    def getTrack(self, filter):
        params = filter.createParameters()
        stream = self._ws.post("track", "", urllib.urlencode(params, True))
        try:
            parser = MbXmlParser()
            return parser.parse(stream).getTrack()
        except ParseError, e:
            raise ResponseError(str(e), e)



lib/python/musicdns/cache.py
"""
Functions to cache the fingerprints and puids.
"""

import os, shelve
from os.path import *

import musicdns


class MusicDNSCache(object):

    def __init__(self, cachefn=None):
        if cachefn is None:
            cachefn = join(os.environ['HOME'], '.musicdns_cache')
        self.cachefn = cachefn
        self.dbm = shelve.open(self.cachefn, 'c')
    
    def getpuid(self, fn, musicdns_key):
        fn = abspath(fn)

        # Lookup puid from the fingerprint.
        try:
            duration, fingerprint, puid = self.dbm[fn]
        except KeyError:
            duration = None
            fingerprint = None
            puid = None

        if fingerprint is None or duration is None:

            # Get the fingerprint and duration.
            try:
                duration, fingerprint, puid = self.dbm[fn]
            except KeyError:
                fingerprint, duration = musicdns.create_fingerprint(fn)
                self.dbm[fn] = duration, fingerprint, puid
                self.dbm.sync()

        if puid is None:
            assert duration is not None
            puid = musicdns.lookup_fingerprint(fingerprint, duration, musicdns_key)
            self.dbm[fn] = duration, fingerprint, puid
            self.dbm.sync()

        return puid, duration



lib/python/musicdns/__init__.py
# -*- coding: utf-8 -*-
#
# pyofa, a wrapper around the MusicDNS libofa and service.
# Copyright (C) 2007 Martin Blais (modifications)
# Copyright (C) 2006 Lukáš Lalinský (original code)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

import sys, os.path, logging, urllib, re

try:
    from xml.etree import ElementTree
    from xml.etree.ElementTree import iterparse
    from xml.etree.ElementPath import Path
except ImportError:
    from elementtree import ElementTree
    from elementtree.ElementTree import iterparse
    from elementtree.ElementPath import Path

from mutagen import File


__version__ = '1.0.3' # 2008-10-06

__all__ = ('initialize', 'finalize',
           'create_fingerprint', 'lookup_fingerprint')


_plugins = ["avcodec", "directshow", "quicktime", "gstreamer"]
_decoders = []

def import_decoders():
    for name in _plugins:
        try:
            decoder = getattr(__import__('musicdns.%s' % name), name)
            _decoders.append(decoder)

            # Note: the reason we're not calling 'init' here is that it may be
            # lengthy, and we want to give the change to multithreaded programs
            # to delegate this to a background thread in order to speed up
            # startup.
        except ImportError, e:
            pass
    if not _decoders:
        logging.warning(
            "No decoders found! Fingerprinting will be disabled.")

def initialize():
    "You must call this before your use the decoders."
    for decoder in _decoders:
        decoder.init()

def finalize():
    "You must call this after you're done using the decoders."
    for decoder in _decoders:
        decoder.done()

try:
    import ofa
except ImportError:
    logging.warning("ofa python module not found! Fingerprinting will be disabled.")
    ofa = None

try:
    import_decoders()
except ImportError:
    logging.warning("Decoder modules not found! Fingerprinting will be disabled.")
    ofa = None

def create_fingerprint(filename):
    """Compute a fingerprint from an open audio file.
    Returns a pair of (fingerprint, duration-in-millis).
    If the file cannot be decoded, a IOError is raised."""

    # Make sure the filename is encoded properly to be consumed by the decoders.
    filename = encode_filename(filename)

    # Try to decode the audio file.
    result = None
    for decoder in _decoders:
        logging.debug("Decoding using %r...", decoder.__name__)
        try:
            result = decoder.decode(filename)
        except Exception, e:
            continue
    if not result:
        raise IOError("Could not decode file '%s' with any of the available decoders: %s." %
                      (filename, ','.join(x.__name__.split('.')[-1] for x in _decoders)))

    # Create the fingerprint.
    #
    # Note: the computed duration is typically incorrect. Use mutagen instead to
    # get the duration. If we could fix this in ffmpeg, we would not need the
    # dependency on mutagen.
    buffer, samples, sample_rate, stereo, _duration = result
    mutagenfile = File(filename)
    if not mutagenfile:
	# This is what should happen, actually:
        #  raise Warning, "Mutagen does not support file %r,
        #                   cannot determine duration"%filename
	duration = -1
    else:
	duration = mutagenfile.info.length * 1000

    fingerprint = ofa.create_print(buffer, samples, sample_rate, stereo)
    return fingerprint, int(duration)

def encode_filename(filename):
    """Encode unicode strings to filesystem encoding."""
    if isinstance(filename, unicode):
        if os.path.supports_unicode_filenames:
            return filename
        else:
            return filename.encode(_io_encoding, 'replace')
    else:
        return filename


musicdns_host = 'ofa.musicdns.org'
musicdns_port = 80

def lookup_fingerprint(fingerprint, duration, musicdns_key, **opt):
    """Given the fingerprint of an audio file, lookup the PUID from the MusicDNS
    service, synchronously."""

    req = '/ofa/1/track'
    url = 'http://%s:%d%s' % (musicdns_host, musicdns_port, req)
    postargs = dict(

        # Identification.
        cid=musicdns_key,
        cvr="pyofa/%s" % __version__,
        
        # The given fingerprint.
        fpt=fingerprint,

        # These are required by the license agreement, to help fill out the
        # MusicDNS database.
        art=opt.pop('art', ''),
        ttl='02-Track_02',
        alb=opt.pop('alb', ''),
        tnm=opt.pop('tnm', ''),  # track no.
        gnr=opt.pop('gnr', ''),
        yrr=opt.pop('yrr', ''),
        brt=opt.pop('brt', ''),
        fmt='', ##MPEG-1 Layer 3',
        dur=str(duration),

        # Return the metadata?
        rmd='0',

        # FIXME: What are those? Found from Picard, not in protocol docs.
        ## rmt='0',
        ## lkt='1',
        )

    data = urllib.urlencode(postargs)
    f = urllib.urlopen(url, data)

    # Parse the response.
    tree = ElementTree.parse(f)
    sanitize_tree(tree)
    el = tree.find('//puid')

    if el is not None:
        puid = el.attrib['id']
    else:
        puid = None

    return puid

    
def sanitize_tree(tree):
    """Remove all XML namespaces from the element tree. This is an inconvenience
    due to the effbot's obsessiveness in maintaining the XML namespaces
    throughout the tags. We prefer BeautifulSoup, which does away with this
    nonsense, but since etree comes with Python, we'll settle for this kludge
    instead in the name of avoiding dependencies.
    """
    for el in tree.getiterator():
        el.tag = re.sub('{.*}', '', el.tag)




lib/python/musicdns/avcodec.c
/*
 * Picard, the next-generation MusicBrainz tagger
 * Copyright (C) 2006 Lukáš Lalinský
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

/* avcodec/avformat-based audio decoder for Picard */

#ifdef _MSC_VER
#define INT64_C(val) val##i64
#define inline __inline
#endif

#include <avcodec.h>
#include <avformat.h>
#include <Python.h>

#ifdef _WIN32

#include <string.h>
#include <avio.h>
#include <fcntl.h>
#include <windows.h>

static int
ufile_open(URLContext *h, const char *filename, int flags)
{
    int access;
    int fd;
    int size;
    const char *ptr;
    wchar_t *w_filename, *w_ptr;
    char *ansi_filename;

    /* skip "ufile:" */
    filename += 6;
    ptr = filename;
    w_filename = malloc(strlen(filename));
    w_ptr = w_filename;
    while (*ptr)
    {
        char a = (*ptr++) - 0x20;
        char b = (*ptr++) - 0x20;
        char c = (*ptr++) - 0x20;
        char d = (*ptr++) - 0x20;
    	  *w_ptr = a | (b << 4) | (c << 8) | (d << 12);
    	  if (*w_ptr == 0)
					break;
				w_ptr++;				   
		}
		*w_ptr = 0;

    if (flags & URL_RDWR) {
        access = O_CREAT | O_TRUNC | O_RDWR;
    } else if (flags & URL_WRONLY) {
        access = O_CREAT | O_TRUNC | O_WRONLY;
    } else {
        access = O_RDONLY;
    }
    access |= O_BINARY;

    if (GetVersion() < 0x80000000) {
        fd = _wopen(w_filename, access, 0666);
    }
    else {
        fd = -1;
        size = wcslen(w_filename) + 2;
        ansi_filename = malloc(size);
        if (ansi_filename) { 
            if (WideCharToMultiByte(CP_ACP, 0, w_filename, -1, ansi_filename, size, NULL, NULL) > 0) {
	  	          fd = open(ansi_filename, access, 0666);
						}
            free(ansi_filename);
				}
    }

    free(w_filename);

    if (fd < 0)
        return AVERROR(ENOENT);
    h->priv_data = (void *)(size_t)fd;
    return 0;
}

static int
ufile_read(URLContext *h, unsigned char *buf, int size)
{
    int fd = (size_t)h->priv_data;
    return read(fd, buf, size);
}

static int
ufile_write(URLContext *h, unsigned char *buf, int size)
{
    int fd = (size_t)h->priv_data;
    return write(fd, buf, size);
}

static offset_t
ufile_seek(URLContext *h, offset_t pos, int whence)
{
    int fd = (size_t)h->priv_data;
    return lseek(fd, pos, whence);
}

static int
ufile_close(URLContext *h)
{
    int fd = (size_t)h->priv_data;
    return _close(fd);
}

URLProtocol ufile_protocol = {
    "ufile",
    ufile_open,
    ufile_read,
    ufile_write,
    ufile_seek,
    ufile_close,
};

#endif

static PyObject *
init(PyObject *self, PyObject *args)
{
    av_register_all();
#ifdef _WIN32
    register_protocol(&ufile_protocol);
#endif
    Py_RETURN_NONE;
}

static PyObject *
done(PyObject *self, PyObject *args)
{
    Py_RETURN_NONE;
}

static PyObject *
decode(PyObject *self, PyObject *args)
{
    AVFormatContext *format_context;
    AVCodecContext *codec_context;
    AVCodec *codec;
    PyObject *filename;
    AVPacket packet;
    unsigned int i;
    int buffer_size, channels, sample_rate, size, len, output_size;
    uint8_t *buffer, *buffer_ptr, *data;
    PyThreadState *_save;
    int duration;

#ifdef _WIN32
    Py_ssize_t w_length;
    wchar_t *w_filename, *w_ptr;
    char *e_filename, *e_ptr;

    if (!PyArg_ParseTuple(args, "U", &filename))
        return NULL;

    /* get the original filename as wchar_t* */
    w_length = PyUnicode_GetSize(filename) + 1;
    w_filename = malloc(w_length * sizeof(wchar_t));
    if (!w_filename)
        return NULL;
    memset(w_filename, 0, w_length * sizeof(wchar_t));
    PyUnicode_AsWideChar((PyUnicodeObject *)filename, w_filename, w_length - 1);
    
    /* 'encode' the filename, so we can pass it as char* */
    e_filename = malloc(w_length * sizeof(wchar_t) * 2 + w_length + 7);
    if (!e_filename)
        return NULL;
    strcpy(e_filename, "ufile:");
    w_ptr = w_filename;
    e_ptr = e_filename + 6;
    while (*w_ptr) {
        *e_ptr++ = 0x20 + ((*w_ptr >>  0) & 0x0F);
        *e_ptr++ = 0x20 + ((*w_ptr >>  4) & 0x0F);
        *e_ptr++ = 0x20 + ((*w_ptr >>  8) & 0x0F);
        *e_ptr++ = 0x20 + ((*w_ptr >> 12) & 0x0F);
        w_ptr++;
    }
    *e_ptr++ = 0x20;
    *e_ptr++ = 0x20;
    *e_ptr++ = 0x20;
    *e_ptr++ = 0x20;
    /* copy ASCII filename to the end for extension-based format detection */		
    w_ptr = w_filename;
    while (*w_ptr) {
        *e_ptr++ = (*w_ptr++) & 0xFF;
    }
    *e_ptr = 0;
		
    Py_UNBLOCK_THREADS;
    if (av_open_input_file(&format_context, e_filename, NULL, 0, NULL) != 0) {
        Py_BLOCK_THREADS;
        free(e_filename);
        free(w_filename);
        PyErr_SetString(PyExc_Exception, "Couldn't open the file.");
        return NULL;
    }

    free(e_filename);
    free(w_filename);
#else
    if (!PyArg_ParseTuple(args, "S", &filename))
        return NULL;

    Py_UNBLOCK_THREADS;
    if (av_open_input_file(&format_context, PyString_AS_STRING(filename), NULL, 0, NULL) != 0) {
        Py_BLOCK_THREADS;
        PyErr_SetString(PyExc_Exception, "Couldn't open the file.");
        return NULL;
    }
#endif

    if (av_find_stream_info(format_context) < 0) {
        Py_BLOCK_THREADS;
        PyErr_SetString(PyExc_Exception, "Couldn't find stream information in the file.");
        return NULL;
    }

#ifndef NDEBUG
    dump_format(format_context, 0, PyString_AS_STRING(filename), 0);
#endif

    codec_context = NULL;
    for (i = 0; i < format_context->nb_streams; i++) {
        codec_context = (AVCodecContext *)format_context->streams[i]->codec;
        if (codec_context && codec_context->codec_type == CODEC_TYPE_AUDIO)
            break;
    }
    if (codec_context == NULL) {
        Py_BLOCK_THREADS;
        PyErr_SetString(PyExc_Exception, "Couldn't find any audio stream in the file.");
        return NULL;
    }

    codec = avcodec_find_decoder(codec_context->codec_id);
    if (codec == NULL) {
        Py_BLOCK_THREADS;
        PyErr_SetString(PyExc_Exception, "Unknown codec.");
        return NULL;
    }

    if (avcodec_open(codec_context, codec) < 0) {
        Py_BLOCK_THREADS;
        PyErr_SetString(PyExc_Exception, "Couldn't open the codec.");
        return NULL;
    }

    channels = codec_context->channels;
    sample_rate = codec_context->sample_rate;

    buffer_size = 135 * channels * sample_rate * 2;
    buffer = (uint8_t *)malloc(buffer_size + AVCODEC_MAX_AUDIO_FRAME_SIZE);
    buffer_ptr = buffer;
    memset(buffer, 0, buffer_size);

    while (buffer_size > 0) {
        if (av_read_frame(format_context, &packet) < 0)
            break;

        size = packet.size;
        data = packet.data;

        while (size > 0) {
            len = avcodec_decode_audio(codec_context, (int16_t *)buffer_ptr, &output_size, data, size);

            if (len < 0)
                break;

            size -= len;
            data += len;

            if (output_size <= 0)
                continue;

            buffer_ptr += output_size;
            buffer_size -= output_size;
            if (buffer_size <= 0)
                break;
        }

        if (packet.data)
            av_free_packet(&packet);
    }

    if (codec_context)
        avcodec_close(codec_context);

    if (format_context)
        av_close_input_file(format_context);

    Py_BLOCK_THREADS;

    /* Compute the duration of the entire file (in millis) */
    /* FIXME: If we can fix getting the duration from ffmpeg, we can get rid
       of the mutagen dependency. */
    if (format_context->duration != AV_NOPTS_VALUE) {
        duration = format_context->duration / AV_TIME_BASE * 1000;
    } else {
        duration = 0;
    }
    /* printf("Duration: %f\n", duration); */
    /* FIXME: this is incorrect. */

    return Py_BuildValue("(N,i,i,i,i)",
        PyCObject_FromVoidPtr(buffer, free),
        (buffer_ptr - buffer) / 2,
        sample_rate,
        channels == 2 ? 1 : 0,
        duration);
}

static PyMethodDef avcodec_methods[] = {
    {"init", init, METH_VARARGS, ""},
    {"done", done, METH_VARARGS, ""},
    {"decode", decode, METH_VARARGS, ""},
    {NULL, NULL, 0, NULL}
};

PyMODINIT_FUNC
initavcodec(void)
{
    (void)Py_InitModule("avcodec", avcodec_methods);
}
